File: src/main/java/no/novari/flyt/configuration/model/configuration/dtos/ObjectMappingDto.java
package no.novari.flyt.configuration.model.configuration.dtos;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.extern.jackson.Jacksonized;
import no.novari.flyt.configuration.validation.constraints.AtLeastOneChild;
import no.novari.flyt.configuration.validation.constraints.UniqueChildrenKeys;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@AtLeastOneChild
@UniqueChildrenKeys
@Getter
@Builder
@EqualsAndHashCode
@Jacksonized
public class ObjectMappingDto {

    public ObjectMappingDto(
            Map<String, ValueMappingDto> valueMappingPerKey,
            Map<String, CollectionMappingDto<ValueMappingDto>> valueCollectionMappingPerKey,
            Map<String, ObjectMappingDto> objectMappingPerKey,
            Map<String, CollectionMappingDto<ObjectMappingDto>> objectCollectionMappingPerKey
    ) {
        this.valueMappingPerKey = Optional.ofNullable(valueMappingPerKey).orElse(new HashMap<>());
        this.valueCollectionMappingPerKey = Optional.ofNullable(valueCollectionMappingPerKey).orElse(new HashMap<>());
        this.objectMappingPerKey = Optional.ofNullable(objectMappingPerKey).orElse(new HashMap<>());
        this.objectCollectionMappingPerKey = Optional.ofNullable(objectCollectionMappingPerKey).orElse(new HashMap<>());
    }

    private final Map<String, @Valid @NotNull ValueMappingDto> valueMappingPerKey;
    private final Map<String, @Valid @NotNull CollectionMappingDto<ValueMappingDto>> valueCollectionMappingPerKey;
    private final Map<String, @Valid @NotNull ObjectMappingDto> objectMappingPerKey;
    private final Map<String, @Valid @NotNull CollectionMappingDto<ObjectMappingDto>> objectCollectionMappingPerKey;

}

File: src/main/kotlin/no/novari/flyt/configuration/model/configuration/dtos/ObjectMappingDto.kt
package no.novari.flyt.configuration.model.configuration.dtos

import jakarta.validation.Valid
import jakarta.validation.constraints.NotNull
import no.novari.flyt.configuration.validation.constraints.AtLeastOneChild
import no.novari.flyt.configuration.validation.constraints.UniqueChildrenKeys

@AtLeastOneChild
@UniqueChildrenKeys
data class ObjectMappingDto(
    @field:Valid
    var valueMappingPerKey: MutableMap<String, @NotNull ValueMappingDto> = mutableMapOf(),
    @field:Valid
    var valueCollectionMappingPerKey: MutableMap<String, @NotNull CollectionMappingDto<ValueMappingDto>> =
        mutableMapOf(),
    @field:Valid
    var objectMappingPerKey: MutableMap<String, @NotNull ObjectMappingDto> = mutableMapOf(),
    @field:Valid
    var objectCollectionMappingPerKey: MutableMap<String, @NotNull CollectionMappingDto<ObjectMappingDto>> =
        mutableMapOf(),
) {
    companion object {
        @JvmStatic
        fun builder(): Builder = Builder()
    }

    class Builder {
        private var valueMappingPerKey: MutableMap<String, ValueMappingDto> = mutableMapOf()
        private var valueCollectionMappingPerKey: MutableMap<String, CollectionMappingDto<ValueMappingDto>> =
            mutableMapOf()
        private var objectMappingPerKey: MutableMap<String, ObjectMappingDto> = mutableMapOf()
        private var objectCollectionMappingPerKey: MutableMap<String, CollectionMappingDto<ObjectMappingDto>> =
            mutableMapOf()

        fun valueMappingPerKey(valueMappingPerKey: Map<String, ValueMappingDto>?) =
            apply { this.valueMappingPerKey = valueMappingPerKey?.toMutableMap() ?: mutableMapOf() }

        fun valueCollectionMappingPerKey(
            valueCollectionMappingPerKey: Map<String, CollectionMappingDto<ValueMappingDto>>?,
        ) = apply {
            this.valueCollectionMappingPerKey = valueCollectionMappingPerKey?.toMutableMap() ?: mutableMapOf()
        }

        fun objectMappingPerKey(objectMappingPerKey: Map<String, ObjectMappingDto>?) =
            apply { this.objectMappingPerKey = objectMappingPerKey?.toMutableMap() ?: mutableMapOf() }

        fun objectCollectionMappingPerKey(
            objectCollectionMappingPerKey: Map<String, CollectionMappingDto<ObjectMappingDto>>?,
        ) = apply {
            this.objectCollectionMappingPerKey = objectCollectionMappingPerKey?.toMutableMap() ?: mutableMapOf()
        }

        fun build(): ObjectMappingDto =
            ObjectMappingDto(
                valueMappingPerKey = valueMappingPerKey,
                valueCollectionMappingPerKey = valueCollectionMappingPerKey,
                objectMappingPerKey = objectMappingPerKey,
                objectCollectionMappingPerKey = objectCollectionMappingPerKey,
            )
    }
}
